processor/src/trace/virtual_bus/sum_check/prover/mod.rs

```diff
@@ -0,0 +1,289 @@
+///
+///         s_i(X_i) := \sum_{(x_{i + 1},\cdots, x_{\nu - 1})
+///                                  w(r_0,\cdots, r_{i - 1}, X_i, x_{i + 1}, \cdots, x_{\nu - 1}).
+///
+///     b. The Verifier checks that s_{i - 1}(r_{i - 1}) = s_{i}(0) + s_{i}(1) rejecting if not.
+///     
+///     c. The Verifier samples a random challenge `r_i ∈ 𝔽` and sends it to the Prover.
+///
+/// 5. The Verifier now queries each of the oracles behind the commitments i.e., `f_i` at
+/// `(r_0, \cdots , r_{\nu - 1})` to get u_i = f_i(r_0, \cdots , r_{\nu - 1}).
+/// The Verifier then accepts if and only if:
+///
+///         s_{\nu - 1}(r_{\nu - 1}) = g(u_0, \cdots , u_{\nu - 1})
+///
+/// A few remarks:
+///
+/// 1. The degree bound on `g` implies that each of the `s_i` polynomials is a univariate polynomial
+/// of degree at most `d`. Thus, the Prover in each round sends `d + 1` values, either
+/// the coefficients or the evaluations of `s_i`.
+///
+/// 2. The Prover has each `f_i` in its evaluation form over the hyper-cube \{0 , 1\}^{2^{\nu}}.
+///
+/// 3. An optimization is for the Prover to not send `s_i(0)` as it can be recoverd from the current
+/// reduced claim s_{i - 1}(r_{i - 1}) using the relation s_{i}(0) = s_{i}(1) - s_{i - 1}(r_{i - 1}).
+/// This also means that the Verifier can skip point 4.b.
+pub struct SumCheckProver<B, E, P, C, H>
+where
+    B: StarkField,
+    E: FieldElement<BaseField = B>,
+    C: RandomCoin<Hasher = H, BaseField = B>,
+    H: ElementHasher<BaseField = B>,
+{
+    composition_poly: P,
+    eval_domain: EvaluationDomain<E>,
+    _challenger: PhantomData<C>,
+}
+
+impl<B, E, P, C, H> SumCheckProver<B, E, P, C, H>
+where
+    B: StarkField,
+    E: FieldElement<BaseField = B>,
+    P: CompositionPolynomial<E>,
+    C: RandomCoin<Hasher = H, BaseField = B>,
+    H: ElementHasher<BaseField = B>,
+{
+    /// Constructs a new [SumCheckProver] given a multivariate composition polynomial.
+    /// The multivariate composition polynomial corresponds to the `g` polynomial in the
+    /// description of the [SumCheckProver] struct.
+    pub fn new(composition_poly: P) -> Self {
+        let max_degree = composition_poly.max_degree();
+        let eval_domain = EvaluationDomain::new(max_degree);
+
+        Self {
+            composition_poly,
+            eval_domain,
+            _challenger: PhantomData,
+        }
+    }
+
+    /// Given an initial claim `claim`, a mutable vector of multi-linear polynomials `mls` and
+    /// a number of rounds `num_rounds`, computes `num_rounds` iterations of the sum-check protocol
+    /// starting from claim `claim`.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - No multi-linears were provided.
+    /// - Number of rounds is zero or is greater than the number of variables of the multilinears.
+    /// - The provided multi-linears have different arities.
+    pub fn prove(
+        &self,
+        claim: E,
+        mls: &mut [MultiLinear<E>],
```

Comment:
We have a name for `g()`: the "composition polynomial". Could we also have a name for the `f_i`s, that's more descriptive than "multilinears"?

Reply:
We could use: multi-linear extension polynomials but I am not sure this is more descriptive